garage/tf/algos/dqn.py

```diff
@@ -0,0 +1,206 @@
+
+                q_best_masked = (1.0 - self.done_t_ph) * future_best_q_val
+                # if done, it's just reward
+                # else reward + discount * future_best_q_val
+                target_q_values = self.reward_t_ph + self.discount * q_best_masked  # noqa: E501
+
+                td_error = q_selected - tf.stop_gradient(target_q_values)
+                loss = huber_loss(td_error)
+                self._loss = tf.reduce_mean(loss)
+
+            with tf.name_scope('optimize_ops'):
+                optimizer = self.qf_optimizer(self.qf_lr)
+                if self.grad_norm_clipping is not None:
+                    gradients = optimizer.compute_gradients(
+                        self._loss, var_list=self.qf.get_trainable_vars())
+                    for i, (grad, var) in enumerate(gradients):
+                        if grad is not None:
+                            gradients[i] = (tf.clip_by_norm(
+                                grad, self.grad_norm_clipping), var)
+                    self._optimize_loss = optimizer.apply_gradients(gradients)
+                else:
+                    self._optimize_loss = optimizer.minimize(
+                        self._loss, var_list=self.qf.get_trainable_vars())
+
+            self._train_qf = tensor_utils.compile_function(
+                inputs=[
+                    self.qf.input, self.action_t_ph, self.reward_t_ph,
+                    self.done_t_ph, self.target_qf.input
+                ],
+                outputs=[self._loss, self._optimize_loss])
+
+    def train_once(self, itr, paths):
+        """Train the algorithm once."""
+        epoch = itr / self.n_epoch_cycles
+
+        self.episode_rewards.extend(paths['undiscounted_returns'])
+        last_average_return = np.mean(self.episode_rewards)
+        for train_itr in range(self.n_train_steps):
+            if self.replay_buffer.n_transitions_stored >= self.min_buffer_size:  # noqa: E501
+                self.evaluate = True
+                qf_loss = self.optimize_policy(epoch, None)
+                self.episode_qf_losses.append(qf_loss)
+
+        if self.evaluate:
+            if itr % self.target_network_update_freq == 0:
+                self._qf_update_ops()
+
+        if itr % self.n_epoch_cycles == 0:
+            if self.evaluate:
+                mean100ep_rewards = round(
+                    np.mean(self.episode_rewards[-100:]), 1)
+                mean100ep_qf_loss = np.mean(self.episode_qf_losses[-100:])
+                tabular.record('Epoch', epoch)
+                tabular.record('AverageReturn', np.mean(self.episode_rewards))
+                tabular.record('StdReturn', np.std(self.episode_rewards))
+                tabular.record('Episode100RewardMean', mean100ep_rewards)
+                tabular.record('{}/Episode100LossMean'.format(self.qf.name),
+                               mean100ep_qf_loss)
+        return last_average_return
+
+    @overrides
+    def get_itr_snapshot(self, itr):
```

Comment:
i think Keren's latest PR will force you to make this class pickleable, so you might as well do that now.

Reply:
It is pickleable now.